### PR TITLE
feat: improved objective function in sampling compressor

### PR DESCRIPTION
**Pull request by lwwmanning**
This PR does several things:
- adds a ton of benchmarks for decompression throughput
- don't allow the patches from bitpacking to be bitpacked with patches
- ALPRD is only called by ALP

But the headliner is that we now add a (configurable) overhead of 64 bytes per descendant array, which penalizes additional levels of cascading compression that don't do much. 

The `compress_noci` benchmark shows  ~10-30% higher compression throughput, and 10-70% higher decompression throughput, with file sizes +/- 3% compared to `develop`.

### CODE REVIEW

**Comment by robert3005** [comment]
Can you separate the last two as separate changes?

File: encodings/runend/src/compress.rs [hunk 1]
```diff
@@ -20,19 +21,21 @@ pub fn runend_encode(array: &PrimitiveArray) -> (PrimitiveArray, PrimitiveArray)
...
-        compressed_values.statistics().set(Stat::IsConstant, false.into());
         compressed_values.statistics().set(Stat::RunCount, compressed_values.len().into());
         array.statistics().get(Stat::Min).map(|s| compressed_values.statistics().set(Stat::Min, s));
         array.statistics().get(Stat::Max).map(|s| compressed_values.statistics().set(Stat::Max, s));
+        array.statistics().get(Stat::IsConstant).map(|s| compressed_values.statistics().set(Stat::IsConstant, s));
         array.statistics().get(Stat::IsSorted).map(|s| compressed_values.statistics().set(Stat::IsSorted, s));
         array.statistics().get(Stat::IsStrictSorted).map(|s| compressed_values.statistics().set(Stat::IsStrictSorted, s));
 
         let compressed_ends = PrimitiveArray::from(ends);
+        compressed_ends.statistics().set(Stat::IsConstant, (compressed_ends.len() > 1).into());
```
Comment (robert3005) [inline encodings/runend/src/compress.rs]: inverse of this?

File: encodings/runend/src/compress.rs [hunk 2]
```diff
@@ -20,19 +21,21 @@ pub fn runend_encode(array: &PrimitiveArray) -> (PrimitiveArray, PrimitiveArray)
...
         compressed_ends.statistics().set(Stat::IsSorted, true.into());
         compressed_ends.statistics().set(Stat::IsStrictSorted, true.into());
-        compressed_ends.statistics().set(Stat::IsConstant, false.into());
-        compressed_ends.statistics().set(Stat::Max, array.len().into());
-        compressed_ends.statistics().set(Stat::RunCount, compressed_ends.len().into());
+        if (compressed_ends.len() > 0) {
```
Comment (robert3005) [inline encodings/runend/src/compress.rs]: Freshly created ends array will have Min of 0 and Max of the array.len(). You don't need to get scalars

File: encodings/runend/src/compress.rs [hunk 3]
```diff
@@ -20,19 +21,21 @@ pub fn runend_encode(array: &PrimitiveArray) -> (PrimitiveArray, PrimitiveArray)
...
         array.statistics().get(Stat::Min).map(|s| compressed_values.statistics().set(Stat::Min, s));
         array.statistics().get(Stat::Max).map(|s| compressed_values.statistics().set(Stat::Max, s));
+        array.statistics().get(Stat::IsConstant).map(|s| compressed_values.statistics().set(Stat::IsConstant, s));
         array.statistics().get(Stat::IsSorted).map(|s| compressed_values.statistics().set(Stat::IsSorted, s));
         array.statistics().get(Stat::IsStrictSorted).map(|s| compressed_values.statistics().set(Stat::IsStrictSorted, s));
 
         let compressed_ends = PrimitiveArray::from(ends);
+        compressed_ends.statistics().set(Stat::IsConstant, (compressed_ends.len() > 1).into());
         compressed_ends.statistics().set(Stat::IsSorted, true.into());
         compressed_ends.statistics().set(Stat::IsStrictSorted, true.into());
```
Comment (robert3005) [inline encodings/runend/src/compress.rs]: I realize now that this is only true if there's no nulls

File: vortex-sampling-compressor/src/compressors/alp_rd.rs [hunk 4]
```diff
@@ -27,6 +27,10 @@ impl EncodingCompressor for ALPRDCompressor {
         ALPRD::ID.as_ref()
     }
 
+    fn decompression_seconds_per_gb(&self) -> f64 {
+        0.39
```
Comment (robert3005) [inline vortex-sampling-compressor/src/compressors/alp_rd.rs]: Can you maybe make this constants in the file so it's easier to know what to tune?

File: vortex-sampling-compressor/src/lib.rs [hunk 5]
```diff
@@ -52,16 +50,74 @@ lazy_static! {
...
+        &FoRCompressor,
+        &SparseCompressor,
+        &ZigZagCompressor,
+    ];
+}
+
+#[derive(Debug, Clone)]
+pub struct ScanPerfConfig {
+    /// bytes per second of throughput
+    bytes_per_ms: f64,
```
Comment (a10y) [inline vortex-sampling-compressor/src/lib.rs]: comment disagrees with var name

**Comment by lwwmanning** [comment]
note: as of commit [45560d4](https://github.com/spiraldb/vortex/pull/1000/commits/45560d4ddb82ffd401178fc4cdd42ddf3a8d3d06), we get 10-25% improved compression throughput with roughly equal file sizes

`BENCH_VORTEX_RATIOS='.*' cargo bench --bench compress_noci -- --baseline develop --test`

where baseline develop is against commit [4aa30c0](https://github.com/spiraldb/vortex/pull/1000/commits/4aa30c0d7ff508b008334213e053b6c9485c00a0)

this is still using the MinSize strategy, albeit now estimating metadata size rather than just data buffer size.

File: pyvortex/src/array.rs [hunk 6]
```diff
@@ -208,19 +208,7 @@ impl PyArray {
     ///         buffer: 1 B
     ///     <BLANKLINE>
     ///
-    /// Compressed arrays use more complex encodings:
```
Comment (robert3005) [inline pyvortex/src/array.rs:211]: Why was this removed?

File: vortex-sampling-compressor/src/lib.rs [hunk 7]
```diff
@@ -173,6 +200,13 @@ impl<'a> SamplingCompressor<'a> {
         cloned
     }
 
+    pub fn only(&self, compressors: &[CompressorRef<'a>]) -> Self {
```
Comment (robert3005) [inline vortex-sampling-compressor/src/lib.rs]: Why? This is like a convoluted version of new

File: vortex-sampling-compressor/src/constants.rs [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#![allow(dead_code)]
+pub mod depth {
```
Comment (robert3005) [inline vortex-sampling-compressor/src/constants.rs]: Don't get why this module is called depth where the value it's storing are cost. Not sure you need additional level of nesting here

File: encodings/fastlanes/src/bitpacking/compress.rs [hunk 9]
```diff
@@ -304,6 +304,27 @@ pub unsafe fn unpack_single_primitive<T: NativePType + BitPacking>(
...
+fn min_patchless_bit_width(bit_width_freq: &[usize]) -> VortexResult<usize> {
+    if bit_width_freq.is_empty() {
+        vortex_bail!("Empty bit width frequency!");
+    }
+    Ok(bit_width_freq
+        .iter()
+        .enumerate()
+        .filter_map(|(bw, count)| (*count > 0).then_some(bw))
+        .max()
+        .unwrap_or_default())
```
Comment (robert3005) [inline encodings/fastlanes/src/bitpacking/compress.rs:325]: potentially other way of doing it is enumerate().rev().drop_while().next() but not sure it's as intuitive to other people as it is to me

File: vortex-sampling-compressor/src/compressors/alp.rs [hunk 10]
```diff
@@ -53,6 +60,7 @@ impl EncodingCompressor for ALPCompressor {
             .map(|p| {
                 ctx.auxiliary("patches")
                     .excluding(self)
+                    .including(&ALPRDCompressor)
```
Comment (robert3005) [inline vortex-sampling-compressor/src/compressors/alp.rs:63]: Seems a bit weird that you would end up here and have it excluded

File: bench-vortex/benches/compress_noci.rs [hunk 11]
```diff
@@ -102,11 +102,32 @@ fn benchmark_compress<T: criterion::measurement::Measurement, F, U>(
...
     group.bench_function(format!("{} compression", bench_name), |b| {
         b.iter_with_large_drop(|| {
-            let compressed = black_box(compressor.compress(uncompressed.as_ref(), None)).unwrap();
+            let compressed = black_box(compressor.compress(uncompressed.as_ref(), None).unwrap());
             compressed_size = compressed.nbytes();
         });
     });
 
+    let compressed = compressor.compress(uncompressed.as_ref(), None).unwrap();
+    let path = format!("{:#?}", compressed.path());
```
Comment (robert3005) [inline bench-vortex/benches/compress_noci.rs]: just get yourself a method to get the path and array at the same time

File: bench-vortex/benches/compress_noci.rs [hunk 12]
```diff
@@ -102,11 +102,32 @@ fn benchmark_compress<T: criterion::measurement::Measurement, F, U>(
...
+
+    let compressed_array = compressed.into_array();
+    //println!("compression tree: {}", compressed_array.tree_display());
```
Comment (robert3005) [inline bench-vortex/benches/compress_noci.rs]: remove this?

File: encodings/dict/benches/dict_compress.rs [hunk 13]
```diff
@@ -49,16 +50,54 @@ fn dict_encode_varbin(arr: &VarBinArray) -> usize {
...
         b.iter(|| black_box(dict_encode_varbin(&varbin_arr)));
     });
 }
 
-criterion_group!(benches, dict_encode);
+fn dict_decode(c: &mut Criterion) {
+    let mut group = c.benchmark_group("dict_decode");
+
+    let primitive_arr = gen_primitive_dict(1_000_000, 0.00005);
+    let (codes, values) = match_each_native_ptype!(primitive_arr.ptype(), |$P| {
```
Comment (robert3005) [inline encodings/dict/benches/dict_compress.rs]: There's a method in dict encoding that does the type switch for you

Comment (robert3005) [reply encodings/dict/benches/dict_compress.rs]: `dict_encode_primitive`?

File: encodings/dict/benches/dict_compress.rs [hunk 14]
```diff
@@ -49,16 +50,54 @@ fn dict_encode_varbin(arr: &VarBinArray) -> usize {
...
+    let varbin_arr = gen_varbin_dict(1_000_000, 0.00005);
+    let (codes, values) = vortex_dict::dict_encode_varbin(&varbin_arr);
+    group.throughput(Throughput::Bytes(varbin_arr.nbytes() as u64));
+    group.bench_function("dict_decode_varbin", |b| {
+        b.iter(|| {
+            black_box(
+                take(values.clone(), codes.clone())
+                    .unwrap()
+                    .into_canonical()
+                    .unwrap(),
```
Comment (robert3005) [inline encodings/dict/benches/dict_compress.rs]: you can create a full dict array and then do `into_canonical`

File: bench-vortex/benches/compressor_throughput.rs [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+#[global_allocator]
+static GLOBAL: MiMalloc = MiMalloc;
+
+fn primitive(c: &mut Criterion) {
+    let mut group = c.benchmark_group("primitive-decompression");
+    let num_values = u16::MAX as u64;
+    group.throughput(Throughput::Bytes(num_values * 4));
+
+    let int_array = PrimitiveArray::from_vec(
+        (0..num_values).map(|i| i as u32 + 100).collect_vec(),
```
Comment (robert3005) [inline bench-vortex/benches/compressor_throughput.rs]: Is the idea that the numbers here are supposed to be linear? Maybe sample from a distribution or just use random number generator

**Review by robert3005 — CHANGES_REQUESTED** [review]
I think there's couple weird artifacts

File: pyvortex/src/array.rs (hunk 6 above)
Comment (lwwmanning) [reply pyvortex/src/array.rs:211]: Because it's super brittle... any change to the compressor can break this example in doctests

File: vortex-sampling-compressor/src/lib.rs (hunk 7 above)
Comment (lwwmanning) [reply vortex-sampling-compressor/src/lib.rs]: except that it doesn't clear `disabled_compressors`, so we can disable stuff further up the tree (e.g., when already used) but be explicit about what the allowed set is further down the tree

File: vortex-sampling-compressor/src/constants.rs (hunk 8 above)
Comment (lwwmanning) [reply vortex-sampling-compressor/src/constants.rs]: Because the config is `max_cost` and the fn on the `EncodingCompressor` trait is `cost`, but it's semantically measuring `depth` (and that's the term from BtrBlocks, we just twiddle it a bit because e.g., some of our encodings are structural, and we don't have another way yet to steer away from e.g., delta-of-deltas)

Happy to take suggestion on how to consolidate the naming; it represents a hard constraint to constrain the search space away from arbitrarily deep compression trees

Comment (robert3005) [reply vortex-sampling-compressor/src/constants.rs]: I’m in favour of getting rid of additional module. Either the usage site provides context or the source code comments. You wouldn’t have constants with same name in different module

File: vortex-sampling-compressor/src/compressors/alp.rs (hunk 10 above)
Comment (lwwmanning) [reply vortex-sampling-compressor/src/compressors/alp.rs:63]: It's not excluded, rather it isn't *included* in the DEFAULT_COMPRESSOR set anymore. This is to say, we only try ALPRD on the ALP patches, rather than as a top-level encoding

Comment (robert3005) [reply vortex-sampling-compressor/src/compressors/alp.rs:63]: I think you still want top level alprd, you still can have top level real doubles. Albeit in tpch there’s probably only one column like that

Comment (lwwmanning) [reply vortex-sampling-compressor/src/compressors/alp.rs:63]: I think we need a way to express top-level preferences. I'd only want to take ALPRD as a top-level compression scheme if ALP failed (e.g., basically all patches)

Comment (robert3005) [reply vortex-sampling-compressor/src/compressors/alp.rs:63]: Right so you can try alprd in alp compressor and then potentially return alprd compressor from alp compressor. I think since we return comperssorref independently of the compressed array you can see play games with implementation

File: vortex-sampling-compressor/src/lib.rs (hunk 7 above)
Comment (robert3005) [reply vortex-sampling-compressor/src/lib.rs]: This method is never called. I think we should remove it and probably also change semantics of including. It's weird that you can include something and it not be included because it's been disabled earlier. At the same time I think if we see usages of this method there's likely a better way of doing things.

File: vortex-sampling-compressor/src/compressors/alp.rs (hunk 10 above)
Comment (lwwmanning) [reply vortex-sampling-compressor/src/compressors/alp.rs:63]: no advantage to doing that, I think. Just reverted for now

**Comment by robert3005** [comment]
We should still get rid of the Compressor::only function. Once we have a use case happy to reconsider

**Review by robert3005 — APPROVED** [review]
